### PR TITLE
EZEE-2661: Allowed translation filtering when publishing version

### PR DESCRIPTION
**Pull request by ViniTou**
| Question           | Answer
| ------------------ | ------------------
| **JIRA issue**     | [EZEE-2661](https://jira.ez.no/browse/EZEE-2661)
| **Bug/Improvement**| yes
| **New feature**    | yes
| **Target version** | `7.5` (`2.5.1`)
| **BC breaks**      | no
| **Tests pass**     | yes
| **Doc needed**     | yes

As a part of fix, we decided to add possibility to publish specific translations only.
As it seems more clearer now, I have decided to move previous logic from Handler inside ContentService and resolve it on API level.

**TODO**:
- [x] Implement feature / fix a bug.
- [x] Implement tests.
- [x] Fix new code according to Coding Standards (`$ composer fix-cs`).
- [x] Ask for Code Review.

### CODE REVIEW

**Comment by pspanja** [comment]
This really looks like a hack on top of a system that doesn't support versioning per language. Because of that it comes with a number of weird edge cases. The first one that came to my mind - what happens when you edit a Content with 2 languages, you change the 1st language but then proceed to publish the 2nd language?

It would be easy to implement this on top of the Public API. What is the reason for adding it there, instead of implementing it for example in the date-based publisher? The similar question goes for https://github.com/ezsystems/ezpublish-kernel/pull/2598

**Comment by ViniTou** [comment]
@pspanja 

>what happens when you edit a Content with 2 translations, you change the 1st translation but then proceed to publish the 2nd translation?

You will end with published version identical to that one without changes to 1st translation. Would you as a user expect different behavior?

>What is the reason for adding it here, instead of implementing it for example in the date-based publisher? 

To have consistent publish action across places.

**Comment by pspanja** [comment]
> You will end with published version identical to that one without changes to 1st translation. Would you as a user expect different behavior?

Not really, it's what I would expect, it's just that this adds another layer of meaning to the whole thing (publishing drafts/fields/translations). I guess it's a matter of how you approach the whole thing, I tend to think Repository should be made as simple and straightforward as possible, and specific behavior should be implemented on top of it. But it's not really my prerogative anymore, so in any case, thanks for the reply 👍

File: eZ/Publish/Core/Repository/ContentService.php [hunk 1]
```diff
@@ -1567,6 +1570,72 @@ public function publishVersion(APIVersionInfo $versionInfo)
...
+                $publishedVersionInfo->languageCodes,
+                $translations
+            );
+        }
+
+        if (empty($languagesToCopy)) {
+            return;
+        }
+
+        $contentType = $this->persistenceHandler->contentTypeHandler()->load($publishedVersionInfo->contentInfo->contentTypeId);
```
Comment (adamwojs) [inline eZ/Publish/Core/Repository/ContentService.php]: IMHO lines 1612 - 1634 might be simplified (?) to 

```php
$contentType = $this->repository->getContentTypeService()->loadContentType(
    $publishedVersionInfo->contentInfo->contentTypeId
);

$updateStruct = $this->newContentUpdateStruct();
foreach ($publishedContent->getFields() as $field) {
	$fieldDefinition = $contentType->getFieldDefinition($field->fieldDefIdentifier);

	if ($fieldDefinition->isTranslatable && in_array($field->languageCode, $languagesToCopy)) {
		$updateStruct->setField($field->fieldDefIdentifier, $field->value, $field->languageCode);
	}
}
```

File: eZ/Publish/Core/Repository/ContentService.php [hunk 2]
```diff
@@ -1567,6 +1570,62 @@ public function publishVersion(APIVersionInfo $versionInfo)
...
+     * @throws \eZ\Publish\API\Repository\Exceptions\ContentValidationException
+     * @throws \eZ\Publish\API\Repository\Exceptions\InvalidArgumentException
+     * @throws \eZ\Publish\API\Repository\Exceptions\NotFoundException
+     * @throws \eZ\Publish\API\Repository\Exceptions\UnauthorizedException
+     */
+    protected function copyTranslationsFromPublishedVersion(APIVersionInfo $versionInfo, array $translations = []): void
+    {
+        $contendId = $versionInfo->contentInfo->id;
+
+        $publishedContent = $this->internalLoadContent($contendId);
```
Comment (alongosz) [inline eZ/Publish/Core/Repository/ContentService.php]: what if there's no published version? Should the last draft be treated the same?

**Comment by micszo** [comment]
Current status:
Flow from https://jira.ez.no/browse/EZEE-2661 - OK.
Flow from https://jira.ez.no/browse/EZP-29694?focusedCommentId=240646&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-240646 - NOK.

**Comment by lserwatka** [comment]
@ViniTou could you look into this test results?

**Pull request merged**